articles/ros_node_idl.md

```diff
@@ -0,0 +1,185 @@
+
+### Other motivations
+
+Outside of security, there are several fascinating possibilities unlocked by having such an interface.
+For example, consider how this could impact [ROS 2 launch][launch_ros].
+It would be able to statically (i.e. before running anything) determine if parameter names or remappings are incorrect, among other similar sanity checks.
+Another example of the usefulness of having a static interface is the ability to create graphical tools for putting a ROS system together.
+Yet another example would be an additional feature in `ros2 pkg create` that would allow a developer to hand it an IDL and have it generate scaffolding for a node with that interface.
+
+These examples are only a subset of use-cases made possible by such an interface.
+It's clear that this is useful well beyond security.
+
+## Package interface
+
+How do upstream packages specify their interface requirements?
+Through a high-level description of all the actions, parameters, services and topics, provided or required, by each node within the package.
+
+The package interface is defined in a separate [XML][xml_wiki] file and exported from the `package.xml` using [REP 149's export mechanism][rep149_export], thereby avoiding pollution of the package manifest.
+The interface may cover only a subset of nodes in a package, as long as the nodes that _are_ covered are done so completely.
+
+Here is an example IDL for a package containing two nodes:
+
+{% include_relative ros2_node_idl/interface_declaration.xml %}
+
+Once an IDL file is written, it is exported from the package manifest:
+
+{% include_relative ros2_node_idl/package.xml %}
+
+### Schema for `package.xml`'s export tag
+
+#### `interface`
+
+This is how the package exports its defined IDL for other tools to consume.
+
+Attributes:
+
+**path**:  Path to XML file containing IDL
+
+Note that the introduction of the `interface` tag within the `export` tag of the package manifest raises a small difficulty with regards to the [REP 149][rep149_export].
+The REP specifies:
+
+> To avoid potential collisions, an export tag should have the same name as the package which is meant to process it. The content of that tag is up to the package to define and use.
+
+Considering the high level abstraction of the IDL, the `interface` tag is not meant for a specific package but rather declares intrinsic properties for anyone to process it.
+However, the keyword is likely solidly descriptive enough to either be accepted as falling under [REP 149][rep149_export] or else to motivate an amendment of the REP.
+
+### IDL Schema
+
+#### `interface`
+
+This tag contains the package interface, made up of a collection of node interfaces.
+Root tag of the IDL file. There must be only one <interface> tag per IDL file.
+
+Attributes:
+- **version**: version of schema in use, allowing for future revisions.
```

Comment:
Why should the `interface` tag itself be versioned and not the `node` tag?

Reply:
The `interface` version correspond to the version of the IDL schema in use, in case it evolves in the future. What should the `node` version represents? The same thing at a finer grain? Is that necessary?

Reply:
I misunderstood the meaning of the version. I was somehow thinking about evolving versions of a nodes interfaces.

Yes, identifying the XML schema in the `interface` tag sounds good - no need for that in the node tag.